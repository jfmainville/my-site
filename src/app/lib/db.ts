"use server";

import { PrismaClient, User, Post, PostStatus } from "@prisma/client";
import { redirect } from "next/navigation";

export type UserSession = {
  user: {
    nickname: string;
    name: string;
    picture: string;
    updated_at: string;
    email: string;
    email_verified: boolean;
    sub: string;
    sid: string;
    accessToken: string;
    accessTokenScope: string;
    accessTokenExpiresAt: string;
    idToken: string;
    token_type: string;
  };
};

const prisma = new PrismaClient();

export async function getPosts() {
  let posts;

  posts = await prisma.post.findMany();

  return posts;
}

export async function getUniquePost(slug: String) {
  let posts;

  if (slug) {
    posts = await prisma.post.findUnique({
      where: {
        slug: slug as string,
      },
    });

    return posts;
  }
}

export async function createPost(
  postTitle: String,
  postCategory: String,
  postThumbnail: String,
  postStatus: String,
  postContent: String,
) {
  let post;

  post = {
    title: postTitle as string,
    thumbnail: postThumbnail as string | null,
    slug: postTitle.replace(" ", "-").toLowerCase() as string,
    content: postContent as string,
    status: (postStatus as PostStatus) || "DRAFT",
    category: postCategory as string,
    // TODO: Need to use the current userId as reference instead of a hardcoded value
    userId: "seed-user-1",
  };
  await prisma.post.create({
    data: post,
  });
  redirect("/admin/post");
}

export async function updatePost(
  postData: Post,
  postTitle: String,
  postThumbnail: String,
  postContent: String,
  postCategory: String,
  postStatus: String,
) {
  let post;

  post = {
    id: postData.id,
    title: postTitle as string,
    thumbnail: postThumbnail as string | null,
    slug: postTitle.replace(" ", "-").toLowerCase() as string,
    content: postContent as string,
    status: postStatus as PostStatus,
    category: postCategory as string,
    // TODO: Need to use the current userId as reference instead of a hardcoded value
    userId: "seed-user-2",
  };
  await prisma.post.update({
    where: {
      id: post.id,
    },
    data: post,
  });
  redirect("/admin/post");
}

export async function deletePost(postId: String) {
  await prisma.post.delete({
    where: {
      id: postId as string,
    },
  });
  redirect("/admin/post");
}

export async function handleUserSession(userSession: UserSession) {
  if (userSession && userSession.user) {
    const userData: User | null = await prisma.user.findUnique({
      where: {
        username: userSession.user?.nickname as string,
      },
    });

    if (!userData) {
      await prisma.user.create({
        data: {
          username: userSession.user.nickname as string,
          email: (userSession.user.email as string) || "",
          firstName: (userSession.user.name?.split(" ")[0] as string) || "",
          lastName: (userSession.user.name?.split(" ")[1] as string) || "",
        },
      });
    }
  }
}
