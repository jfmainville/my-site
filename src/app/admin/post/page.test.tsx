import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import PostPage from "./page";
import { getPosts } from "@/app/lib/db";

jest.mock("../../lib/db");

jest.mock("next/image", () => {
  return function MockImage({ src, alt }: { src: string; alt: string }) {
    return <img src={src} alt={alt} />;
  };
});

describe("<PostPage/> Test Suite", () => {
  it("should render the <PostPage/> page unchanged", async () => {
    const { container } = render(await PostPage());
    expect(container).toMatchSnapshot();
  });

  it("should return existing posts in the <PostPage/> page", async () => {
    (getPosts as jest.Mock).mockResolvedValueOnce([
      {
        id: "0eb6cb4f-2020-4465-bb89-c4081753c71e",
        slug: "test-blog-post-1",
        title: "Test Blog Post 1",
        content: "<h1>Test Blog Post 1</h1>",
        status: "DRAFT",
        category: "Cybersecurity",
        userId: "48eceb63-fc36-426f-85a5-58ccc850eade",
      },
      {
        id: "37314b8f-6d97-4daa-affa-1717864a8e0e",
        slug: "test-blog-post-2",
        title: "Test Blog Post 2",
        content: "<h1>Test Blog Post 2</h1>",
        status: "DRAFT",
        category: "Cybersecurity",
        userId: "830ba444-6d36-401a-95cc-28646bfad2d5",
      },
    ]);
    render(await PostPage());
    const firstBlogPost = screen.getByText("Test Blog Post 1");
    const secondBlogPost = screen.getByText("Test Blog Post 2");
    expect(firstBlogPost).toBeInTheDocument();
    expect(secondBlogPost).toBeInTheDocument();
  });

  it("should return a placeholder if there are no posts in the <AdminPage/> page", async () => {
    render(await PostPage());
    const noPostMessage = screen.getByText("No posts");
    expect(noPostMessage).toBeInTheDocument();
  });
});
