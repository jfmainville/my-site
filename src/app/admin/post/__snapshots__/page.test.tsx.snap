// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostPage/> Test Suite should render the <PostPage/> page unchanged 1`] = `
<div>
  <section
    class="Main"
  >
    <div
      class="Navbar"
    >
      <div>
        Admin
      </div>
      <div>
        Logout
      </div>
    </div>
    <a
      class="Link"
      href="/admin"
    >
      <div
        class="Button"
        style="color: black; background-color: transparent;"
      >
        <div>
          &lt;- Back
        </div>
      </div>
    </a>
    <div
      class="CreateButton"
    >
      <a
        class="Link"
        href="/admin/post/create"
      >
        <div
          class="Button"
        >
          <div>
            Create Post
          </div>
        </div>
      </a>
    </div>
    <div
      class="BlogPost"
    >
      <div>
        Name
      </div>
      <div>
        Status
      </div>
      <div>
        Create Date
      </div>
      <div>
        Update Date
      </div>
      <div>
        No posts
      </div>
    </div>
  </section>
</div>
`;
